### PR TITLE
[Chunk Data Pack Pruner] Add block view index

### PR DESCRIPTION
**Pull request by zhangchiqing**
This PR adds an index for lookup block by view. The index is only added to certified block which has received a QC. 

This index allows us to iterate through each view, and prune data indexed by the block ID.

Note: The method is not available until either next spork or a migration to build the index is completed. 

This PR is based on the following PRs, better to review them before this:
- https://github.com/onflow/flow-go/pull/7452
- https://github.com/onflow/flow-go/pull/7469

### CODE REVIEW

**Review by AlexHentschel — COMMENTED** [review]
I very much agree with implementing a `ByView` indexing and retrieval. Though, I am a bit uncomfortable with the assumptions and undocumented dependencies that the proposed change would introduce:
* the [`QuorumCertificates` populates the ByView index](https://github.com/onflow/flow-go/blob/700b677f7f14db8f0533fe1379a7dac7f2cbc7c4/storage/badger/qcs.go#L36) without ever seeing a block, while the [`Headers`](https://github.com/onflow/flow-go/blob/92ac8d18e6721e6d63d8f6b0326639c43e5ea01c/storage/badger/headers.go#L17) just assumes that if that index exist, a corresponding block has been previously stored. Nowhere does `QuorumCertificates` document that _only_ QCs for known blocks are to be stored. 
* `QuorumCertificates`  and `Headers` are low-level storage abstractions, but they make very specific assumptions how the higher level code (specifically the `Mutator`) is supposed to use them: https://github.com/onflow/flow-go/blob/92ac8d18e6721e6d63d8f6b0326639c43e5ea01c/storage/badger/qcs.go#L35 Also that is not in detail documented as part of the API, so for anyone to not accidentally violate this, they would need to go into he implementation and check, which in my experience engineers don't do. 
   * Just hypothetically, if somebody has the great idea to persist QCs on consensus nodes once they construct them (in case the node crashes) and maybe even do that for QCs pertaining to unknown blocks, the assumption made in `QuorumCertificates` would be void. Its probably not the best example, but it is _an_ example how fragile those assumptions can be. 

So in summary, I am to very comfortable with changing the meaning of  [`QuorumCertificates`](https://github.com/onflow/flow-go/blob/ffc03982ffac7b85c4932b1eb717a87d70b6a2f6/storage/qcs.go#L13) from "storing quorum certificates" to "storing QuorumCertificates _only for known blocks_". Essentially, we are introducing additional dependencies and assumptions between `QuorumCertificates`, `Headers` and `Mutator`. While in each individual case, the coupling is small, over time such couplings accumulate across the code base and make the code very difficult to maintain and to guarantee correctness. A major focus of our work is managing complexity and guaranteeing correctness, if that requires to write a bit more boilerplate code - I think that would be a good tradeoff. 

Some thoughts / suggestions:
* since `Headers` retrieves the blocks by view and assumes that for an indexed view a block exists, it should also populate the by-view index in my opinion (keeping the code that deals with the same thing in one place). 
*  The `Mutator` is the module that has the authority to declare blocks as finalized, sealed, etc ... I think the authority to declare a block as certified also falls within that category (certainly I would discourage the give the storage layer the authority to declare a block certified). I would suggest to make that explicit by having the `Mutator` call a dedicated method `IndexCertifiedBlockByView(flow.Identifier, uint64)` that is provided by the storage abstraction (e.g. `Headers`). 
* In the way I see it [`Blocks`](https://github.com/onflow/flow-go/blob/7e87d5b9c53464cf854003af0f30ceacf3c4f8a1/storage/badger/blocks.go#L16) is an extension of [`Headers`](https://github.com/onflow/flow-go/blob/92ac8d18e6721e6d63d8f6b0326639c43e5ea01c/storage/badger/headers.go#L17). 
   * Therefore, also `Blocks` could offer the `ByView` method 
   * `Blocks` could also offer the `DeclareCertifiedTx` method and delegate that to `IndexCertifiedBlockByView` (?) To me, that would seem very consistent. 

* method `IndexCertifiedBlockByView`  would need to be called by the `Mutator` in the following two places:
  * before [line 368](https://github.com/onflow/flow-go/blob/357925d4b83cbe28b5d8c567de205cd3f5a6b230/state/protocol/badger/mutator.go#L368) 
  * and before [line 392](https://github.com/onflow/flow-go/blob/357925d4b83cbe28b5d8c567de205cd3f5a6b230/state/protocol/badger/mutator.go#L392)

File: state/protocol/badger/mutator.go [hunk 1]
```diff
@@ -365,6 +365,12 @@ func (m *FollowerState) headerExtend(ctx context.Context, candidate *flow.Block,
 				return fmt.Errorf("could not store incorporated qc: %w", err)
 			}
 		} else {
+			// parent is a block that has been received and certified by a QC.
+			err := transaction.WithTx(operation.SkipDuplicates(operation.IndexCertifiedBlockByView(parent.View, qc.BlockID)))(tx)
```
Comment (zhangchiqing) [inline state/protocol/badger/mutator.go:369]: @AlexHentschel I was using the db operation directly, for 2 reasons:

I wanted to add a IndexCertifiedBlockByView method to headers, and simply call m.headers.IndexCertifiedBlockByView(parent) here. But the storage.Headers interface is mostly for read methods, we didn't expose methods like IndexFinalizedBlockByView because we didn't want this method being misused.
For the same reason I also didn't add IndexCertifiedBlockByView method to Blocks, and also, the parent block body doesn't exist in this context, passing a block to blocks.IndexCertifiedBlockByView would require us to make additional query to the parent block body.

**Comment by AlexHentschel** [comment]
regarding your suggestion for file `storage/badger/operation/headers.go`
> we might want to change [`IndexBlockHeight`](https://github.com/onflow/flow-go/blob/c901bc4cba207d61f3477f8a09f34bdb9b937560/storage/badger/operation/headers.go#L19) into `IndexFinalizedBlockByHeight`?

yes, I would appreciate that :-)

**Comment by zhangchiqing** [comment]
>we might want to change [IndexBlockHeight](https://github.com/onflow/flow-go/blob/c901bc4cba207d61f3477f8a09f34bdb9b937560/storage/badger/operation/headers.go#L19) into IndexFinalizedBlockByHeight?

Fixed f383dcbbe6

**Review by AlexHentschel — APPROVED** [review]
Thanks Leo for the great PR. Appreciate your patience with the review time 🙈. 

I have created a documentation PR on aspects that I wondered about while reviewing your PR ... mostly forward-facing thoughts on where the `mutator` code would need to change in the future when moving to pebble. I think the only place where my PR is relevant for your immediate changes is my argument that we don't need to `Skip Duplicates` (my comments [here](https://github.com/onflow/flow-go/pull/6933#discussion_r1974682817) and [here](https://github.com/onflow/flow-go/pull/6933#discussion_r1974682949)). 
I am happy if you merge your PR first, and then we iterate over my documentation - don't feel obligated to include that as part of your PR .... though, don't hesitate to merge mine it into your branch if you feel it would help Jordan with his review 🤷.

**Review by jordanschalm — APPROVED** [review]
Looks good. There are a few documentation changes I didn't understand, but I think those are from @AlexHentschel's [PR](https://github.com/onflow/flow-go/pull/7101).

File: state/protocol/badger/mutator.go [hunk 2]
```diff
@@ -135,6 +135,9 @@ func NewFullConsensusState(
 // CAUTION:
 //   - This function expects that `certifyingQC ` has been validated. (otherwise, the state will be corrupted)
 //   - The parent block must already have been ingested.
+//   - Attempts to extend the state with the _same block concurrently_ are not allowed.
+//     (will not corrupt the state, but may lead to an exception)
+// Orphaned blocks are excepted.
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/mutator.go:140]: If it means "above caution does not apply", I don't think it is accurate. If it means "triggers an exception" I think we should say that instead.

File: storage/badger/operation/headers.go [hunk 3]
```diff
@@ -20,11 +20,22 @@ func IndexBlockHeight(height uint64, blockID flow.Identifier) func(*badger.Txn)
 	return insert(makePrefix(codeHeightToBlock, height), blockID)
 }
 
+// IndexBlockView indexes the view of a block. It should only be called on block that
+// have received QC
+func IndexBlockView(view uint64, blockID flow.Identifier) func(*badger.Txn) error {
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply storage/badger/operation/headers.go]: > `IndexFinalizedBlockByHeight`

Thanks. I like that suggestion. 👍  Included a [todo in issue #7682](https://github.com/onflow/flow-go/issues/7682) (low prio) if we wanted to come back to this.

File: state/protocol/badger/mutator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/mutator.go]: Thanks for catching this. For function [`ParticipantState.Extend`](https://github.com/onflow/flow-go/blob/13c527fe1711eea04277c24f19a7e7f0c651a9c5/state/protocol/badger/mutator.go#L245C10-L245C26) this comment in [line 222](https://github.com/onflow/flow-go/blob/13c527fe1711eea04277c24f19a7e7f0c651a9c5/state/protocol/badger/mutator.go#L222) this is indeed incorrect: https://github.com/onflow/flow-go/blob/13c527fe1711eea04277c24f19a7e7f0c651a9c5/state/protocol/badger/mutator.go#L222  the documentation  states clearly the contrary: https://github.com/onflow/flow-go/blob/13c527fe1711eea04277c24f19a7e7f0c651a9c5/state/protocol/badger/mutator.go#L243

I have checked the implementation (specifically function [`checkOutdatedExtension`](https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L483-L510)) and [documented it in much more detail](https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L483-L510) to avoid similar confusions what "outdated" precisely means. 

Regarding your question what it means that "Orphaned blocks are excepted", it is explained below: https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L142-L151

I think that explanation is much more detailed and clear. Hence in https://github.com/onflow/flow-go/pull/6933/commits/1013553a3277d48ed1c13ee11a497c84cd487df9 I have removed the summary "Orphaned blocks are excepted"

File: state/protocol/badger/mutator.go [hunk 4]
```diff
@@ -201,8 +214,12 @@ func (m *FollowerState) ExtendCertified(ctx context.Context, candidate *flow.Blo
...
 // the validity of the _entire block_ (header and full payload).
 //
-// CAUTION: per convention, the protocol state requires that the candidate's
-// parent has already been ingested. Otherwise, an exception is returned.
+// CAUTION:
+//   - per convention, the protocol state requires that the candidate's
+//     parent has already been ingested. Otherwise, an exception is returned.
+//   - Attempts to extend the state with the _same block concurrently_ are not allowed.
+//     (will not corrupt the state, but may lead to an exception)
+// Orphaned blocks are excepted.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/mutator.go]: same as above: https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L229-L238

the problem with my change is: it is wrong 🤦 . In contrast to [`FollowerState.ExtendCertified`](https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L161), the method [`ParticipantState.Extend`](https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L250) rejects orphaned blocks. This was already documented: https://github.com/onflow/flow-go/blob/357925d4b83cbe28b5d8c567de205cd3f5a6b230/state/protocol/badger/mutator.go#L219 In my in commit https://github.com/onflow/flow-go/pull/6933/commits/1013553a3277d48ed1c13ee11a497c84cd487df9 today, I have documented the inner workings of [`ParticipantState.checkOutdatedExtension`](https://github.com/onflow/flow-go/blob/1013553a3277d48ed1c13ee11a497c84cd487df9/state/protocol/badger/mutator.go#L250) in much more.

[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/mutator.go]: clarified in commit https://github.com/onflow/flow-go/pull/6933/commits/9528bafcf8e5133c15952a490399ac9c4608c478

File: state/protocol/chain_state.go [hunk 5]
```diff
@@ -53,8 +53,17 @@ type FollowerState interface {
 	// CAUTION:
 	//   - This function expects that `qc` has been validated. (otherwise, the state will be corrupted)
 	//   - The parent block must already be stored.
+	//   - Attempts to extend the state with the _same block concurrently_ are not allowed.
+	//     (will not corrupt the state, but may lead to an exception)
 	// Orphaned blocks are excepted.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/chain_state.go]: clarified in commit https://github.com/onflow/flow-go/pull/6933/commits/9528bafcf8e5133c15952a490399ac9c4608c478

File: state/protocol/chain_state.go [hunk 6]
```diff
@@ -79,8 +88,19 @@ type ParticipantState interface {
...
 	// The candidate block must have passed HotStuff validation before being passed to Extend.
 	//
-	// CAUTION: per convention, the protocol state requires that the candidate's
-	// parent has already been ingested. Otherwise, an exception is returned.
+	// CAUTION:
+	//   - per convention, the protocol state requires that the candidate's
+	//     parent has already been ingested. Otherwise, an exception is returned.
+	//   - Attempts to extend the state with the _same block concurrently_ are not allowed.
+	//     (will not corrupt the state, but may lead to an exception)
+	// Orphaned blocks are excepted.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/chain_state.go]: clarified in commit https://github.com/onflow/flow-go/pull/6933/commits/9528bafcf8e5133c15952a490399ac9c4608c478

File: state/protocol/badger/mutator.go [hunk 7]
```diff
@@ -190,6 +200,9 @@ func (m *FollowerState) ExtendCertified(ctx context.Context, candidate *flow.Blo
 
 	// Execute the deferred database operations as one atomic transaction and emit scheduled notifications on success.
 	// The `candidate` block _must be valid_ (otherwise, the state will be corrupted)!
+	// Therefore, they should be identified
+	// additional are just a no-op if they are identified as such early enough by the `checkBlockAlreadyProcessed` - though if the same extension is added concurrently, we might return a
+	// [storage.ErrAlreadyExists] error.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/mutator.go]: thanks. Sorry. These sentence fragments are messed up .. not sure what happened, maybe copy paste gone wrong, maybe stopped writing. Anyway I fixed it as follows: https://github.com/onflow/flow-go/blob/fb40d3ca4e7350d56f446d8cdd43e7dc8d0a831b/state/protocol/badger/mutator.go#L323-L329

File: state/protocol/badger/mutator.go [hunk 8]
```diff
@@ -256,6 +278,14 @@ func (m *ParticipantState) Extend(ctx context.Context, candidate *flow.Block) er
...
 	}
 
+	deferredBlockPersist := deferred.NewDeferredBlockPersist()
+
+	// check if the block header is a valid extension of parent block
+	err = m.headerExtend(ctx, candidate, nil, deferredBlockPersist)
+	if err != nil {
+		return fmt.Errorf("header not compliant with chain state: %w", err)
+	}
+
```
Comment (AlexHentschel) [inline state/protocol/badger/mutator.go:288]: for people reviewing my changes: 
* I have moved `headerExtend` below the calls [`checkBlockAlreadyProcessed`](https://github.com/onflow/flow-go/blob/b4951912cb40210a07018cd719987c2bf5b2de90/state/protocol/badger/mutator.go#L264) and [`checkOutdatedExtension`](https://github.com/onflow/flow-go/blob/b4951912cb40210a07018cd719987c2bf5b2de90/state/protocol/badger/mutator.go#L273)

thereby, we run the sanity checks first, before the more expensive business logic

File: utils/unittest/fixtures.go
Comment (AlexHentschel) [inline utils/unittest/fixtures.go:1]: Regarding the functions `BlockWithParentFixtureAndUniqueView` and `BlockWithParentProtocolStateAndUniqueView`, which you have introduced, note that they can lead to subtle flakey tests (tests running indefinitely and timing out). This is because we generate the child's view by taking the parent view and adding a random number [1,10]:

https://github.com/onflow/flow-go/blob/68910c3622dca0c91ed17c5742a0be53be0e5e43/utils/unittest/fixtures.go#L558-L561

If all of those views are already used, the functions will just keep running! 

I have tried to address this in https://github.com/onflow/flow-go/pull/6933/commits/7e9e63c0bf839afea2f4751a97813c3a6100e134 (and also changed the functions from recursive to loops)

File: state/protocol/badger/mutator_test.go [hunk 9]
```diff
@@ -167,21 +175,22 @@ func TestSealedIndex(t *testing.T) {
...
 
+		viewIndex := make(map[uint64]struct{})
 		// build a chain:
 		// G <- B1 <- B2 (resultB1) <- B3 <- B4 (resultB2, resultB3) <- B5 (sealB1) <- B6 (sealB2, sealB3) <- B7
 		// test that when B4 is finalized, can only find seal for G
 		// 					 when B5 is finalized, can find seal for B1
 		//					 when B7 is finalized, can find seals for B2, B3
 
 		// block 1
-		b1 := unittest.BlockWithParentFixture(rootHeader)
```
Comment (AlexHentschel) [inline state/protocol/badger/mutator_test.go]: @zhangchiqing I don't understand why you have made this change. `BlockWithParentFixture` creates a child, whose view is strictly larger then the parent's ([eventually calling into this code here](https://github.com/onflow/flow-go/blob/7e9e63c0bf839afea2f4751a97813c3a6100e134/utils/unittest/fixtures.go#L558-L560)). 

Note that we are specifically using the fixture `BlockWithParentFixture` here that creates a child of a specified parent. In my opinion, the fixtures should be producing sensible data. If `BlockWithParentFixture` were to produce a child with the same view, that would be a bud in the fixture. Nothing to be fixed in the test! 

Note the important difference:
* `BlockWithParentFixtureAndUniqueView` or `BlockWithParentProtocolStateAndUniqueView` are only needed if we want to test with different forks. Only there, generating different blocks with the same view is even possible.

I prefer tests to be as concise as possible. I think the usage of `BlockWithParentFixtureAndUniqueView` or `BlockWithParentProtocolStateAndUniqueView`  in _some_ of the tests adds unnecessary complexity. Furthermore, the usage of those methods is incomplete in several tests making it more confusing (e.g. test [`TestSealedIndex`](https://github.com/onflow/flow-go/blob/7e9e63c0bf839afea2f4751a97813c3a6100e134/state/protocol/badger/mutator_test.go#L171) mixes the usage of `BlockWithParentFixtureAndUniqueView` and [`BlockWithParentProtocolState`](https://github.com/onflow/flow-go/blob/7e9e63c0bf839afea2f4751a97813c3a6100e134/state/protocol/badger/mutator_test.go#L201-L204))


For the _tests_ that **create only a single linear fork**, I reverted your usage of `BlockWithParentFixtureAndUniqueView` or `BlockWithParentProtocolStateAndUniqueView` in https://github.com/onflow/flow-go/pull/6933/commits/d8dbf0f5d0a0107e6c054636fcc0d8405395aa7b. Please let me know if you disagree.

File: utils/unittest/fixtures.go [hunk 10]
```diff
@@ -379,6 +379,32 @@ func BlockWithParentFixture(parent *flow.Header) *flow.Block {
...
 
+// BlockWithParentFixtureAndUniqueView creates a child block of the given parent.
+// We provide a set of views that are _not_ allowed to be used for the new block. A typical usage
+// scenario is to create blocks of different forks, without accidentally creating two blocks with
+// the same view.
+// CAUTION:
+//   - modifies the set `forbiddenViews` by adding the view of the newly created block.
+//   - To generate the child's view, we randomly select a small increment and add it to the
+//     parent's view. If the set of views covers all possible increments, this function will panic
+func BlockWithParentFixtureAndUniqueView(parent *flow.Header, forbiddenViews map[uint64]struct{}) *flow.Block {
```
Comment (AlexHentschel) [inline utils/unittest/fixtures.go]: The term "Fixture" in our context applies to the objects we are creating. The function "BlockWithParentFixture" is supposed to mean "create a block fixture with the given parent".

I feel the term "Fixture" is overall not very helpful and just bloats method names. Below, you introduce the method `BlockWithParentProtocolStateAndUniqueView` -- without the word "Fixture" in the name. I like that a lot better. 

In commit https://github.com/onflow/flow-go/pull/6933/commits/d8dbf0f5d0a0107e6c054636fcc0d8405395aa7b, I have refactored this method here by removing the word "Fixture" from the method name.

File: state/protocol/badger/mutator_test.go [hunk 11]
```diff
@@ -1160,13 +1184,14 @@ func TestExtendConflictingEpochEvents(t *testing.T) {
 		result, _, err := rootSnapshot.SealedResult()
 		require.NoError(t, err)
 
+		viewIndex := make(map[uint64]struct{})
 		// add two conflicting blocks for each service event to reference
-		block1 := unittest.BlockWithParentFixture(head)
+		block1 := unittest.BlockWithParentFixtureAndUniqueView(head, viewIndex)
```
Comment (AlexHentschel) [inline state/protocol/badger/mutator_test.go]: In this test, I agree with your usage of `BlockWithParentFixtureAndUniqueView`, because we create two conflicting forks. However, note that the test still has edge cases, where ure-usage of view numbers is possible:

https://github.com/onflow/flow-go/blob/7e9e63c0bf839afea2f4751a97813c3a6100e134/state/protocol/badger/mutator_test.go#L1271-L1279

I have fixed this in https://github.com/onflow/flow-go/pull/6933/commits/d8dbf0f5d0a0107e6c054636fcc0d8405395aa7b

File: state/protocol/badger/snapshot_test.go [hunk 12]
```diff
@@ -666,15 +669,15 @@ func TestSealingSegment(t *testing.T) {
 	// where DefaultTransactionExpiry = 600
 	t.Run("test extra blocks contain exactly DefaultTransactionExpiry number of blocks below the sealed block", func(t *testing.T) {
 		util.RunWithFollowerProtocolState(t, rootSnapshot, func(db *badger.DB, state *bprotocol.FollowerState) {
-			root := unittest.BlockWithParentFixture(head)
+			viewIndex := make(map[uint64]struct{})
+			root := unittest.BlockWithParentFixtureAndUniqueView(head, viewIndex)
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot_test.go]: similarly here, we are just creating a single long fork. As `BlockWithParentFixture` creates a child with view that is always larger than the parent, there is no risk of view collisions -- unless I am overlooking something. 

I reverted the usage of `BlockWithParentFixtureAndUniqueView` in commit https://github.com/onflow/flow-go/pull/6933/commits/d8dbf0f5d0a0107e6c054636fcc0d8405395aa7b to not clutter up the test with unnecessary logic -- please let me know if i am misunderstanding something here.

File: state/protocol/badger/mutator_test.go [hunk 13]
```diff
@@ -545,44 +553,52 @@ func TestExtendMissingParent(t *testing.T) {
...
-		require.NoError(t, err)
+	blockC := unittest.BlockWithParentFixture(blockB.Header) // creates child with height one larger, but view possibly much larger
+	blockC.SetPayload(unittest.PayloadFixture(unittest.WithProtocolStateID(rootProtocolStateID)))
+	blockC.Header.Height = blockB.Header.Height
+	//blockC.Header.View = blockB.Header.View + 1
 
-		// create another block with the same height and view, that is coming after
-		extend.Header.ParentID = extend.Header.ID()
-		extend.Header.Height = 1
-		extend.Header.View = 2
```
Comment (AlexHentschel) [inline state/protocol/badger/mutator_test.go]: In the prior code, we were creating block `extend` and then added it to the protocol state. After adding it, we modified the block in place. This caused problems because the `state` is potentially caching the block. 

I fixed this in commit https://github.com/onflow/flow-go/pull/6933/commits/561c90e66c6446f8aae79a7651e07e45c11c26c7 by independently creating two blocks.

File: state/protocol/badger/snapshot_test.go [hunk 14]
```diff
@@ -674,7 +674,6 @@ func TestSealingSegment(t *testing.T) {
 			parent := root
 			for i := 0; i < flow.DefaultTransactionExpiry+1; i++ {
 				next := unittest.BlockWithParentProtocolState(parent)
-				next.Header.View = next.Header.Height + 1 // set view so we are still in the same epoch
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot_test.go:677]: By default, `BlockWithParentProtocolState` creates a child block, whose view can be up to 10 bigger than its parent. If we are unlucky and randomly select unfavourable view increments, the resulting view of the last generated block might fall beyond the epoch. The comment explicitly explains this. By removing this, we are probably making the test flakey. 

I fixed this in commit https://github.com/onflow/flow-go/pull/6933/commits/54001f71af279d443e81f61c730efb67aeea76be

File: state/protocol/badger/snapshot_test.go [hunk 15]
```diff
@@ -887,13 +892,18 @@ func TestSealingSegment_FailureCases(t *testing.T) {
 
 	// SCENARIO 2b: An orphaned block is chosen as head; at this height a block other than the orphaned has been finalized.
 	t.Run("sealing segment from orphaned block", func(t *testing.T) {
+		// In this test, we create two conflicting forks. To prevent accidentally creating byzantine scenarios, where
+		// multiple blocks have the same view, we keep track of used views and ensure that each new block has a unique view.
+		usedViews := make(map[uint64]struct{})
+		usedViews[sporkRoot.View] = struct{}{}
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot_test.go:898]: Here, we are creating conflicting forks, i.e. there is the potential for randomly generating descendants with the same view number. 

In https://github.com/onflow/flow-go/pull/6933/commits/54001f71af279d443e81f61c730efb67aeea76be I updated the test to use your methods `BlockWithParentFixtureAndUniqueView` or `BlockWithParentProtocolStateAndUniqueView` 🎉

File: storage/operation/multi_dbstore.go [hunk 16]
```diff
@@ -32,3 +32,13 @@ func (b *multiDBStore) WithReaderBatchWriter(fn func(storage.ReaderBatchWriter)
 func (b *multiDBStore) NewBatch() storage.Batch {
 	return b.rwStore.NewBatch()
 }
+
+func (b *multiDBStore) Close() error {
```
Comment (AlexHentschel) [inline storage/operation/multi_dbstore.go]: same comment as on PR https://github.com/onflow/flow-go/pull/7469

> ⚠️ (?) even if one db errors (e.g. failed disk), shouldn't we still try to close the other (potentially still open on another disk), before we return?  
> 
> I think this is a good use of `multierror.Error`? 

It doesn't really matter where this is resolved ... just wanting to make sure it doesn't fall through the cracks.

File: storage/badger/operation/prefix.go [hunk 17]
```diff
@@ -10,106 +10,13 @@ const (
...
-
-	// internal failure information that should be preserved across restarts
-	codeExecutionFork                   = 254
-	codeEpochEmergencyFallbackTriggered = 255
+	codeEpochSetup       = 61 // EpochSetup service event, keyed by ID
+	codeEpochCommit      = 62 // EpochCommit service event, keyed by ID
+	codeBeaconPrivateKey = 63 // BeaconPrivateKey, keyed by epoch counter
+	// _                      = 64 // [DEPRECATED] flag that the DKG for an epoch has been started, used in protocol version v1
+	codeDKGEndState = 65 // [DEPRECATED] flag for DKG end state, used in protocol version v1
+	codeDKGState    = 66 // current state of Recoverable Random Beacon State Machine for given epoch
```
Comment (AlexHentschel) [inline storage/badger/operation/prefix.go]: Here is my understanding, why we still have badger storage codes:
* the secrets database is still badger
* it stores DKG related stuff (incl. the private key)

For now, storage codes that exist in both files have the same value. However, I am worried that  in the **future**, the **same storage constant might have different values** for badger and pebble (e.g. value changed for one but not the other) -- and if we accidentally confuse the storage codes from badger and pebble we might corrupt our data. 

I think the following convention would be good to reduce probability of subtle bugs:
* any storage constant (name) is defined _either_ in pebble _or_ in badger, but not in both! 

That way, pebble code trying to use the badger storage constants (or vice versa) won't compile. 

I have implemented the following changes in commit https://github.com/onflow/flow-go/pull/6933/commits/b3561f90ab7b3ac08c21d494b85c46e957f89707 . Please take a look to make sure I didn't screw up anything (its late already 😅 ):

-  removed [`codeEpochSetup` and `codeEpochCommit`](https://github.com/onflow/flow-go/blob/ad322dfa9b5f0ea8853bf473b889851a48191725/storage/badger/operation/prefix.go#L14-L15) from `storage/badger/operation/prefix.go` as they are used the protocol database but not the Secrets database.
- the variables for storage codes `64` and `65` can be deprecated, as neither is used anymore _anywhere_.  The respective lines now read: https://github.com/onflow/flow-go/blob/b3561f90ab7b3ac08c21d494b85c46e957f89707/storage/operation/prefix.go#L77-L78
- To prevent accidental usage of variables from the wrong file, I have I removed the declaration  of `codeBeaconPrivateKey` (storage code `63`) from `storage/operation/prefix.go`. The respective line now reads: https://github.com/onflow/flow-go/blob/b3561f90ab7b3ac08c21d494b85c46e957f89707/storage/operation/prefix.go#L79
- storage code `66` was previously entirely removed from `storage/operation/prefix.go`.  I have included the following dummy to emphasize that this storage code is used in a different database now: https://github.com/onflow/flow-go/blob/b3561f90ab7b3ac08c21d494b85c46e957f89707/storage/operation/prefix.go#L79

File: storage/operation/prefix.go [hunk 18]
```diff
@@ -11,8 +11,8 @@ import (
 const (
 
 	// codes for special database markers
-	// codeMax    = 1 // deprecated
-	codeDBType = 2 // specifies a database type
+	_ = 1 // DEPRECATED: previously used for badger to denote the max length of the storage codes in units of bytes
+	_ = 2 // DEPRECATED: previously used to differentiate the protocol database from the secrets database; now the former is pebble and the latter is badger
```
Comment (AlexHentschel) [inline storage/operation/prefix.go:15]: those storage codes were only used by badger (file `storage/badger/operation/prefix.go`). I have removed the constant's declaration here (file `storage/operation/prefix.go`) so that we can't accidentally use a equally named storage code (with potentially different value) from teh wrong database. 

Please feel free to revert, if you disagree.

File: utils/unittest/fixtures.go
Comment (zhangchiqing) [reply utils/unittest/fixtures.go:1]: Good catch! Thanks for the fix.

File: state/protocol/badger/mutator_test.go (hunk 9 above)
Comment (zhangchiqing) [reply state/protocol/badger/mutator_test.go]: makes sense. Thanks for the update.

File: storage/operation/prefix.go (hunk 18 above)
Comment (zhangchiqing) [reply storage/operation/prefix.go:15]: Yup, makes sense.

File: storage/operation/prefix.go [hunk 19]
```diff
@@ -72,9 +73,10 @@ const (
 	// codes related to protocol level information
 	codeEpochSetup         = 61 // EpochSetup service event, keyed by ID
 	codeEpochCommit        = 62 // EpochCommit service event, keyed by ID
-	codeBeaconPrivateKey   = 63 // BeaconPrivateKey, keyed by epoch counter
-	codeDKGStarted         = 64 // flag that the DKG for an epoch has been started
-	codeDKGEnded           = 65 // flag that the DKG for an epoch has ended (stores end state)
+	_                      = 63 // USED BY SECRETS DATABASE: BeaconPrivateKey, keyed by epoch counter
+	_                      = 64 // DEPRECATED: flag that the DKG for an epoch has been started
+	_                      = 65 // DEPRECATED: flag that the DKG for an epoch has ended (stores end state)
+	_                      = 66 // USED BY SECRETS DATABASE: current state of Recoverable Random Beacon State Machine for given epoch
```
Comment (zhangchiqing) [inline storage/operation/prefix.go:79]: +1
